core/src/main/java/org/apache/brooklyn/core/effector/composite/LoopEffector.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.brooklyn.core.effector.composite;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.brooklyn.api.effector.Effector;
+import org.apache.brooklyn.api.entity.Entity;
+import org.apache.brooklyn.config.ConfigKey;
+import org.apache.brooklyn.core.config.ConfigKeys;
+import org.apache.brooklyn.core.effector.AddEffector;
+import org.apache.brooklyn.core.effector.EffectorBody;
+import org.apache.brooklyn.core.effector.Effectors.EffectorBuilder;
+import org.apache.brooklyn.core.entity.EntityInitializers;
+import org.apache.brooklyn.util.core.config.ConfigBag;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.Beta;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+@Beta
+public class LoopEffector extends AbstractCompositeEffector {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LoopEffector.class);
+
+    public static final ConfigKey<String> INPUT = ConfigKeys.newStringConfigKey(
+            "input",
+            "Loop input parameter");
+
+    public static final ConfigKey<Object> LOOP = ConfigKeys.newConfigKey(
+            Object.class,
+            "loop",
+            "Effector details for the loop effector");
+
+    public LoopEffector(ConfigBag params) {
+        super(newEffectorBuilder(params).build());
+    }
+
+    public LoopEffector(Map<?, ?> params) {
+        this(ConfigBag.newInstance(params));
+    }
+
+    public static EffectorBuilder<List> newEffectorBuilder(ConfigBag params) {
+        EffectorBuilder<List> eff = AddEffector.newEffectorBuilder(List.class, params);
+        EffectorBody<List> body = new Body(eff.buildAbstract(), params);
+        eff.impl(body);
+        return eff;
+    }
+
+    protected static class Body extends AbstractCompositeEffector.Body<List> {
+
+        public Body(Effector<?> eff, ConfigBag params) {
+            super(eff, params);
+            Preconditions.checkNotNull(params.getAllConfigRaw().get(LOOP.getName()), "Effector names must be supplied when defining this effector");
```

Comment:
Should this be `Preconditions.checkNotNull(params.get(LOOP), "loop must be supplied when defining a loop effector: %s", params)`?